File: p043.java
/* 
 * Solution to Project Euler problem 43
 * By Nayuki Minase
 * 
 * http://nayuki.eigenstate.org/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public final class p043 {
	
	public static void main(String[] args) {
		System.out.println(new p043().run());
	}
	
	
	public String run() {
		int[] divisibilityTests = {2, 3, 5, 7, 11, 13, 17};
		
		int[] digits = new int[10];
		for (int i = 0; i < digits.length; i++)
			digits[i] = i;
		
		long sum = 0;
		outer:
		do {
			for (int i = 0; i < divisibilityTests.length; i++) {
				if (toInteger(digits, i + 1, 3) % divisibilityTests[i] != 0)
					continue outer;
			}
			System.out.println(toInteger(digits, 0, digits.length));
			sum += toInteger(digits, 0, digits.length);
		} while (Library.nextPermutation(digits));
		return Long.toString(sum);
	}
	
	
	private static long toInteger(int[] digits, int off, int len) {
		long result = 0;
		for (int i = off; i < off + len; i++)
			result = result * 10 + digits[i];
		return result;
	}
	
}

File: p043.py
# 
# Solution to Project Euler problem 43
# by Project Nayuki
# 
# http://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 


def compute():
	ans = 0
	num = list(range(10))
	while True:
		if is_substring_divisible(num):
			ans += int("".join(map(str, num)))
		if not next_permutation(num):
			break
	return str(ans)


DIVISIBILITY_TESTS = [2, 3, 5, 7, 11, 13, 17]

def is_substring_divisible(num):
	for i in range(7):
		x = num[i + 1] * 100 + num[i + 2] * 10 + num[i + 3]
		if x % DIVISIBILITY_TESTS[i] != 0:
			return False
	return True


def next_permutation(arr):
	i = len(arr) - 1
	while i > 0 and arr[i - 1] >= arr[i]:
		i -= 1
	if i <= 0:
		return False
	j = len(arr) - 1
	while arr[j] <= arr[i - 1]:
		j -= 1
	arr[i - 1], arr[j] = arr[j], arr[i - 1]
	arr[i : ] = arr[len(arr) - 1 : i - 1 : -1]
	return True


if __name__ == "__main__":
	print(compute())
